### PR TITLE
fix: Allow config CRD changes to cause a reconcilliation loop for dependent (but not owned) resources

### PR DESCRIPTION
**Pull request by cliveseldon**
At present changes to SeldonConfig do not cause a reconcile of associated SeldonRuntimes and ServerConfig changes do not cause a reconcile for Servers. This would hamper rollout of changes.

There is not a direct ownership from SeldonConfigs to SeldonRuntimes or ServerConfigs to Servers as SeldonRuntimes and Servers are created later and have a `spec` link to an associated config.

The solution is to add to the controllers watches with custom reconcile e.g.

```
Watches(
			&source.Kind{Type: &mlopsv1alpha1.SeldonConfig{}},
			handler.EnqueueRequestsFromMapFunc(r.mapSeldonRuntimesFromSeldonConfig),
		).
```

where we add 
```
// Find SeldonRuntimes that reference the changes SeldonConfig
func (r *SeldonRuntimeReconciler) mapSeldonRuntimesFromSeldonConfig(obj client.Object) []reconcile.Request {
	logger := log.FromContext(context.Background()).WithName("mapSeldonRuntimesFromSeldonConfig")
	var seldonRuntimes mlopsv1alpha1.SeldonRuntimeList
	if err := r.Client.List(context.Background(), &seldonRuntimes); err != nil {
		logger.Error(err, "error listing seldonRuntimes")
		return nil
	}

	seldonConfig := obj.(*mlopsv1alpha1.SeldonConfig)
	var req []reconcile.Request
	for _, seldonRuntime := range seldonRuntimes.Items {
		if seldonRuntime.Spec.SeldonConfig == seldonConfig.Name {
			req = append(req, reconcile.Request{
				NamespacedName: types.NamespacedName{
					Namespace: seldonRuntime.Namespace,
					Name:      seldonRuntime.Name,
				},
			})
		}
	}
	return req
}
```

### CODE REVIEW

File: operator/controllers/mlops/seldonruntime_controller.go [hunk 1]
```diff
@@ -217,6 +220,30 @@ func (r *SeldonRuntimeReconciler) updateStatus(seldonRuntime *mlopsv1alpha1.Seld
 	return nil
 }
 
+// Find SeldonRuntimes that reference the changes SeldonConfig
+func (r *SeldonRuntimeReconciler) mapSeldonRuntimesFromSeldonConfig(obj client.Object) []reconcile.Request {
+	logger := log.FromContext(context.Background()).WithName("mapSeldonRuntimesFromSeldonConfig")
+	var seldonRuntimes mlopsv1alpha1.SeldonRuntimeList
+	if err := r.Client.List(context.Background(), &seldonRuntimes); err != nil {
+		logger.Error(err, "error listing seldonRuntimes")
+		return nil
```
Comment (sakoush) [inline operator/controllers/mlops/seldonruntime_controller.go:229]: do we want to pass an error to the caller?

File: operator/controllers/mlops/server_controller.go [hunk 2]
```diff
@@ -226,11 +229,39 @@ func (r *ServerReconciler) updateStatus(server *mlopsv1alpha1.Server) error {
 	return nil
 }
 
+// Find Servers that need reconcilliation from a change to a given ServerConfig
+func (r *ServerReconciler) mapServerFromServerConfig(obj client.Object) []reconcile.Request {
+	logger := log.FromContext(context.Background()).WithName("mapServerFromServerConfig")
+	var servers mlopsv1alpha1.ServerList
+	if err := r.Client.List(context.Background(), &servers); err != nil {
+		logger.Error(err, "error listing servers")
+		return nil
```
Comment (sakoush) [inline operator/controllers/mlops/server_controller.go:238]: ditto

**Review by sakoush — APPROVED** [review]
LGTM, I left a minor comment on error handling

It would also be good to add this somewhere in the docs to explain the behaviour for the user.

Also worth adding a follow up ticket to make this configurable.

File: operator/controllers/mlops/seldonruntime_controller.go (hunk 1 above)
Comment (cliveseldon) [reply operator/controllers/mlops/seldonruntime_controller.go:229]: connected to https://github.com/kubernetes-sigs/controller-runtime/issues/2235 so not sure how to handle will need to look further

File: operator/controllers/mlops/server_controller.go (hunk 2 above)
Comment (cliveseldon) [reply operator/controllers/mlops/server_controller.go:238]: https://github.com/kubernetes-sigs/controller-runtime/issues/2235

**Pull request merged**